### PR TITLE
Horizontal Scrolling

### PR DESCRIPTION
**Pull request by bambu**
Added support for horizontal scrolling. The code comes from the original
vim source. Removed the check for `guioptions` as Neovim does not have
support for that setting. Including a test for `<ScrollWheelLeft>` and
`<ScrollWheelRight>` in `mouse_spec.lua`.

This PR is for #3440.

### CODE REVIEW

**Comment by justinmk** [comment]
Lint errors. Also, functions without parameters require (void).

File: src/nvim/normal.c [hunk 1]
```diff
@@ -3611,6 +3625,107 @@ void scroll_redraw(int up, long count)
...
+{
+    char_u  *p;
+    colnr_T col;
+    int   w;
+
+    p = ml_get(lnum);
+    col = 0;
+    if (*p != NUL)
+      for (;;)
+      {
```
Comment (ianks) [inline src/nvim/normal.c]: This if statement is very ambiguous

Comment (bambu) [inline src/nvim/normal.c]: That is the way that it is written in the VIM source. If necessary, I can modify it to be less ambiguous but I tried to leave the code as close as possible to the original.

**Comment by bambu** [comment]
Is there any action needed that is preventing this from being merged?

File: src/nvim/normal.c (hunk 1 above)
Comment (ZyX-I) [inline src/nvim/normal.c]: @bambu “As close as possible” is a bad idea. We have different code style.

File: src/nvim/edit.c [hunk 2]
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
             (long)(curwin->w_botline - curwin->w_topline));
       else
         scroll_redraw(dir, 3L);
+    } else {
+      int val, step = 6;
```
Comment (ZyX-I) [inline src/nvim/edit.c]: Split the definition.

File: src/nvim/edit.c [hunk 3]
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
...
+
+      if (mod_mask & (MOD_MASK_SHIFT | MOD_MASK_CTRL))
+        step = curwin->w_width;
```
Comment (ZyX-I) [inline src/nvim/edit.c]: `{}` over each `if` body.

File: src/nvim/normal.c (hunk 1 above)
Comment (ZyX-I) [inline src/nvim/normal.c]: `{` should be on the same line as `for`.

File: src/nvim/normal.c [hunk 4]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
 }
 
 /*
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Multiline `/**/` comments are not allowed, use `///`. How did linter accept this?

File: src/nvim/normal.c [hunk 5]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+ */
+static linenr_T gui_find_longest_lnum(void)
+{
+    linenr_T ret = 0;
+
+    /* Calculate maximum for horizontal scrollbar.  Check for reasonable
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indentation. `&&` should be on the same column as the very first `curwin`.

File: src/nvim/normal.c [hunk 6]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `{` should be on the same line as `if`, or at least the last `)`

File: src/nvim/edit.c [hunk 7]
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
...
+      val = curwin->w_leftcol + (dir == MSCR_RIGHT ? -step : step);
+      if (val < 0)
+        val = 0;
```
Comment (ZyX-I) [inline src/nvim/edit.c]: `{}`

File: src/nvim/normal.c [hunk 8]
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
       cap->count0 = 3;
       nv_scroll_line(cap);
     }
+  } else {
+    /* Horizontal scroll - only allowed when 'wrap' is disabled */
+    if (!curwin->w_p_wrap) {
+      int val, step = 6;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Split.

File: src/nvim/normal.c [hunk 9]
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
...
+
+      if (mod_mask & (MOD_MASK_SHIFT | MOD_MASK_CTRL))
+          step = curwin->w_width;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}

File: src/nvim/normal.c [hunk 10]
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
...
+
+      val = curwin->w_leftcol + (cap->arg == MSCR_RIGHT ? -step : +step);
+      if (val < 0)
+          val = 0;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}

File: src/nvim/normal.c [hunk 11]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
 }
 
 /*
+ * Return length of line "lnum" for horizontal scrolling.
+ */
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
+    colnr_T col;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Should be moved to the place where col is set to zero.

File: src/nvim/normal.c [hunk 12]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
 }
 
 /*
+ * Return length of line "lnum" for horizontal scrolling.
+ */
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Should be moved to the place where `p` is assigned.

File: src/nvim/normal.c [hunk 13]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    colnr_T col;
+    int   w;
+
+    p = ml_get(lnum);
+    col = 0;
+    if (*p != NUL)
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}

File: src/nvim/normal.c [hunk 14]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      for (;;)
+      {
+          w = chartabsize(p, col);
+          mb_ptr_adv(p);
+          if (*p == NUL)    /* don't count the last character */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}, `//` comment

File: src/nvim/normal.c [hunk 15]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+            break;
+          col += w;
+      }
+    return col;
+}
+
+/*
+ * Find longest visible line number.  If this is not possible (or not desired,
+ * by setting 'h' in "guioptions") then the current line number is returned.
+ */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `///`

File: src/nvim/normal.c [hunk 16]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+static linenr_T gui_find_longest_lnum(void)
+{
+    linenr_T ret = 0;
+
+    /* Calculate maximum for horizontal scrollbar.  Check for reasonable
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `//`

File: src/nvim/normal.c [hunk 17]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    return col;
+}
+
+/*
+ * Find longest visible line number.  If this is not possible (or not desired,
+ * by setting 'h' in "guioptions") then the current line number is returned.
+ */
+static linenr_T gui_find_longest_lnum(void)
+{
+    linenr_T ret = 0;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indentation: should be two spaces.

File: src/nvim/normal.c [hunk 18]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
 }
 
 /*
+ * Return length of line "lnum" for horizontal scrolling.
+ */
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
+    colnr_T col;
+    int   w;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Move to where `w` is assigned. This is absolutely not needed in the function scope.

File: src/nvim/normal.c [hunk 19]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+
+    p = ml_get(lnum);
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indentation.

File: src/nvim/normal.c [hunk 20]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      linenr_T    lnum;
+      colnr_T     n;
+      long      max = 0;
+
+      /* Use maximum of all visible lines.  Remember the lnum of the
+       * longest line, closest to the cursor line.  Used when scrolling
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
+      {
+          n = scroll_line_len(lnum);
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indent.

File: src/nvim/normal.c [hunk 21]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
+      colnr_T     n;
+      long      max = 0;
+
+      /* Use maximum of all visible lines.  Remember the lnum of the
+       * longest line, closest to the cursor line.  Used when scrolling
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Move `{` here.

File: src/nvim/normal.c [hunk 22]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
+      colnr_T     n;
+      long      max = 0;
+
+      /* Use maximum of all visible lines.  Remember the lnum of the
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `//`

File: src/nvim/normal.c [hunk 23]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    linenr_T ret = 0;
+
+    /* Calculate maximum for horizontal scrollbar.  Check for reasonable
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Should be `for (linenr_T lnum =…`, not here.

File: src/nvim/normal.c [hunk 24]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      colnr_T     n;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Should be inside the `for` cycle.

File: src/nvim/normal.c [hunk 25]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
+      {
+          n = scroll_line_len(lnum);
+          if (n > (colnr_T)max)
+          {
+            max = n;
+            ret = lnum;
+          } else if (n == (colnr_T)max
+            && abs((int)(lnum - curwin->w_cursor.lnum))
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indent.

File: src/nvim/normal.c [hunk 26]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+               < abs((int)(ret - curwin->w_cursor.lnum))) {
+              ret = lnum;
+          }
+      }
+    } else {
+      /* Use cursor line only. */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: `//`

File: src/nvim/normal.c [hunk 27]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      ret = curwin->w_cursor.lnum;
+    }
+
+    return ret;
+}
+
+static linenr_T longest_lnum;
+/*
+ * Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+ */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: ///

File: src/nvim/normal.c [hunk 28]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
+{
+    /* no wrapping, no scrolling */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: //

File: src/nvim/normal.c [hunk 29]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    if (curwin->w_p_wrap)
+      return false;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}

File: src/nvim/normal.c [hunk 30]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+
+    if (curwin->w_leftcol == (colnr_T)leftcol)
+      return false;
```
Comment (ZyX-I) [inline src/nvim/normal.c]: {}

File: src/nvim/normal.c [hunk 31]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+
+    curwin->w_leftcol = (colnr_T)leftcol;
+
+    /* When the line of the cursor is too short, move the cursor to the
+     * longest visible line. */
+    if (!virtual_active()
+      && (colnr_T)leftcol > scroll_line_len(curwin->w_cursor.lnum))
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Align after (

File: src/nvim/normal.c [hunk 32]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    {
+      if (compute_longest_lnum)
```
Comment (ZyX-I) [inline src/nvim/normal.c]: if(){

File: src/nvim/normal.c [hunk 33]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+      {
+          curwin->w_cursor.lnum = gui_find_longest_lnum();
+          curwin->w_cursor.col = 0;
+      } else if (longest_lnum >= curwin->w_topline
+        && longest_lnum < curwin->w_botline) {
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Align

File: src/nvim/normal.c [hunk 34]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+        /* Do a sanity check on "longest_lnum", just in case. */
```
Comment (ZyX-I) [inline src/nvim/normal.c]: //

File: src/nvim/normal.c (hunk 28 above)
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indent: 4 in place of 2.

File: src/nvim/normal.c [hunk 35]
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
...
+    char_u  *p;
+    colnr_T col;
+    int   w;
+
+    p = ml_get(lnum);
+    col = 0;
+    if (*p != NUL)
+      for (;;)
+      {
+          w = chartabsize(p, col);
```
Comment (ZyX-I) [inline src/nvim/normal.c]: Wrong indent as far as I see.

**Comment by ZyX-I** [comment]
Linter is close to useless here.

**Comment by christian-schulze** [comment]
Dump.

This would be a really awesome addition to neovim.

File: src/nvim/normal.c [hunk 36]
```diff
@@ -3940,6 +3957,101 @@ void scroll_redraw(int up, long count)
...
+  }
+
+  return ret;
+}
+
+static linenr_T longest_lnum;
+///
+/// Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+///
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
```
Comment (justinmk) [inline src/nvim/normal.c]: This and the other 2 new functions can be moved to `mouse.c` after adding these includes to mouse.c:

    #include "nvim/memline.h"
    #include "nvim/charset.h"

File: src/nvim/normal.c [hunk 37]
```diff
@@ -3940,6 +3957,101 @@ void scroll_redraw(int up, long count)
...
+      }
+      col += w;
+    }
+  }
+  return col;
+}
+
+///
+/// Find longest visible line number.  If this is not possible (or not desired,
+/// by setting 'h' in "guioptions") then the current line number is returned.
```
Comment (justinmk) [inline src/nvim/normal.c]: You removed the check for 'guioptions' so let's remove that sentence from this comment. Also makes sense to remove the `gui_` prefix from the function name now.

File: src/nvim/normal.c [hunk 38]
```diff
@@ -3867,6 +3867,23 @@ static void nv_mousescroll(cmdarg_T *cap)
...
+          step = curwin->w_width;
+      }
+
+      val = curwin->w_leftcol + (cap->arg == MSCR_RIGHT ? -step : +step);
+      if (val < 0) {
+          val = 0;
+      }
+
+      gui_do_horiz_scroll(val, true);
+    }
```
Comment (justinmk) [inline src/nvim/normal.c]: This whole block is identical with the block added in edit.c. Can you move it to a static function in mouse.c, something like `mouse_scroll_horiz`. Then `gui_do_horiz_scroll` can be made static, too.

**Comment by bambu** [comment]
@justinmk Made the requested updates.

**Comment by justinmk** [comment]
@bambu LGTM, thanks for followup! Merged w/ squash.

**Pull request closed without merging**